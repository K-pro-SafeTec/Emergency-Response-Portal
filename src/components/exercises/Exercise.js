import React from 'react';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import AppPage from '../shared/AppPage';
import EntityInfo from '../shared/EntityInfo';
import IconExercise from '../../icons/main-menu/exerciseScenarios.svg';
import { exerciseScenarioById } from '../../dummy-data/exerciseScenario';
import { getExerciseInstanceByKey } from '../../dummy-data/exerciseInstance';


const Exercise = ({ match }) => {
  const instance = getExerciseInstanceByKey(match.params.scenarioId, match.params.date);
  if (instance) {
    const scenario = exerciseScenarioById[match.params.scenarioId];
    return (
      <AppPage title="Øvelser" back="../..">
        <Paper style={{margin: '1em', padding: '1em'}}>
          <EntityInfo Icon={(props) => <img src={IconExercise} alt="Øvelse" {...props} />}>
            <Typography variant="headline">{scenario.scenario}</Typography>
            <Typography variant="subheading">DFU {scenario.dfu}</Typography>
          </EntityInfo>
          <Typography variant="subheading">Scenariobeskrivelse:</Typography>
          <Typography variant="body1" paragraph>{scenario.description}</Typography>
          <Typography variant="subheading">Forberedelser:</Typography>
          <Typography variant="body1" paragraph>{scenario.preparations}</Typography>
          <Typography variant="subheading">Ytelseskrav:</Typography>
          <Typography variant="body1" paragraph>{scenario.performanceRequirements}</Typography>
          <Typography variant="subheading">Parametere:</Typography>
          <ul>
            <li>
              <Typography variant="body1">Vind: {scenario.parameters.wind}</Typography>
            </li>
            <li>
              <Typography variant="body1">Bølger: {scenario.parameters.waves}</Typography>
            </li>
            <li>
              <Typography variant="body1">POB: {scenario.parameters.pob}</Typography>
            </li>
            <li>
              <Typography variant="body1">Pågående aktiviteter: {scenario.parameters.ongoingActivities}</Typography>
            </li>
          </ul>
          <Typography variant="subheading">Fokusområder:</Typography>
          <Typography variant="body1" paragraph>{scenario.focusAreas}</Typography>
          <Typography variant="subheading">Utvikling:</Typography>
          <Typography variant="body1" paragraph>{scenario.development}</Typography>
        </Paper>
      </AppPage>
    );
  }
  return (
    <AppPage title="Øvelser" back="../..">
      <Typography>Fant ikke øvelse</Typography>
    </AppPage>
  );
};

export default Exercise;
