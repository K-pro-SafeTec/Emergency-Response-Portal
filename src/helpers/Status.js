const Status = {
  OK: Symbol("OK"),
  WARNING: Symbol("WARNING"),
  ERROR: Symbol("ERROR"),
};

export default Status;

const severity = {
  [Symbol.OK]: 0,
  [Symbol.WARNING]: 1,
  [Symbol.ERROR]: 2,
};

export function mostSevere(a, b) {
  return severity.get(a) > severity.get(b) ? a : b;
}
