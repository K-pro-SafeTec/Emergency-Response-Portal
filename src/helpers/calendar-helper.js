// Takes javascript date object and returns nice time
function getTimeFormat(date) {
  let hours = date.getHours();
  let minutes = date.getMinutes();
  if (hours < 10) {
    hours = "0" + hours;
  }
  if (minutes < 10) {
    minutes = "0" + minutes;
  }
  return hours + ":" + minutes;
}

function getDateFormatYMD(date) {
  const year = date.getFullYear();
  let month = date.getMonth() + 1;
  if (month < 10) {
    month = "0" + month;
  }
  let dayInMonth = date.getDate();
  if (dayInMonth < 10) {
    dayInMonth = "0" + dayInMonth;
  }
  return year + "-" + month + "-" + dayInMonth
}

// Takes a list of events and a date as objects.
// Returns the events that starts on the given date.
function getEvents(events, date) {
  return events.filter(event => {
    return equalDates(event.start, date)
  });
}

// Takes two javascript date objects as arguments and returns true if the objects fall on the same date.
function equalDates(date1, date2) {
  const dayInMonth1 = date1.getDate();
  const dayInMonth2 = date2.getDate();
  const month1 = date1.getMonth();
  const month2 = date2.getMonth();
  const year1 = date1.getYear();
  const year2 = date2.getYear();
  return dayInMonth1 === dayInMonth2 && month1 === month2 && year1 === year2;
}

// Takes in a javascript date object and returns a nice string.
function getDateFormat(date) {
  const weekdays = ["Søndag", "Mandag", "Tirsdag", "Onsdag", "Torsdag", "Fredag", "Lørdag"];
  const months = ["januar", "februar", "mars", "april", "mai", "juni", "juli", "august", "september", "oktober",
    "november", "desember"];
  const weekday = weekdays[date.getDay()];
  const month = months[date.getMonth()];
  return weekday + " " + date.getDate() + ". " + month + " " + date.getFullYear()
}

// Checks if an event is a valid event
function isValidEvent(date, start, end, participants, type) {
  return isValidDate(date) && isValidPeriod(start, end) && isValidParticipants(participants) &&
      isValidType(type)
}

// Returns true if the input date is on a format that can be parsed to a date
function isValidDate(date) {
  return !isNaN(Date.parse(date));
}

// Returns true if the input time is on the format tt:mm
function isValidTime(time) {
  return /^([01][0-9]|2[0-3]):[0-5][0-9]$/.test(time)
}

function isValidPeriod(start, end) {
  if (!isValidTime(start) || !isValidTime(end)) {
    return false;
  }
  return start <= end;
}

function isValidParticipants(participants) {
  const valids = ['alle', 'beredskapslag', 'beredskapsledelse'];
  return valids.includes(participants)
}

function isValidType(type) {
  const valids = ['øvelse', 'trening', 'table top'];
  return valids.includes(type)
}

function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

function sortEvents(events) {
  events.sort(function(a, b){
    if (a.start - b.start === 0) {
      return a.end - b.end;
    }
    return a.start - b.start;
  });
  return events;
}

// Returns the date of a special weekday some weeks in the future or in the past.
// 0 is Sunday, 1 is Monday etc.
// For instance, running the method getDateRelatedToThisDate(2, 3) will return the date of Wednesday in three weeks.
function getDateRelatedToThisDate(weekdayNumber, numberOfWeeks, hours=0, minutes=0) {
  let date = new Date();
  date.setDate(date.getDate()+ 7 * numberOfWeeks + weekdayNumber - date.getDay());
  date.setHours(hours);
  date.setMinutes(minutes);
  return date;
}

export { getTimeFormat, getDateFormatYMD, getEvents, equalDates, getDateFormat, isValidEvent, isValidParticipants,
  isValidType, capitalizeFirstLetter, sortEvents, getDateRelatedToThisDate }
