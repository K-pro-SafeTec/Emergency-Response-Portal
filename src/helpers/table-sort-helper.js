function desc(a, b, orderBy) {
  if (b[orderBy] < a[orderBy]) {
    return -1;
  }
  if (b[orderBy] > a[orderBy]) {
    return 1;
  }
  return 0;
}

function statusDesc(a, b, competenceNumber) {
  const statuses = ["Symbol(OK)", "Symbol(WARNING)", "Symbol(ERROR)"];
  let indexA = 2;
  let indexB = 2;
  if (a.competence[competenceNumber]) {
    indexA = statuses.indexOf(a.competence[competenceNumber].status.toString());
  }
  if (b.competence[competenceNumber]) {
    indexB = statuses.indexOf(b.competence[competenceNumber].status.toString());
  }
  if (indexB < indexA) {
    return -1;
  }
  if (indexB > indexA) {
    return 1;
  }
  return 0;
}

export function stableSort(array, cmp) {
  const stabilizedThis = array.map((el, index) => [el, index]);
  stabilizedThis.sort((a, b) => {
    const order = cmp(a[0], b[0]);
    if (order !== 0) return order;
    return a[1] - b[1];
  });
  return stabilizedThis.map(el => el[0]);
}

export function getSorting(order, orderBy) {
  const descFunction = typeof orderBy === "number" ? statusDesc : desc;
  return order === "desc"
    ? (a, b) => descFunction(a, b, orderBy)
    : (a, b) => -descFunction(a, b, orderBy);
}
