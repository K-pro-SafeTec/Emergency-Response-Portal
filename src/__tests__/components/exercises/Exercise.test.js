import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom'
import Exercise from "../../../components/exercises/Exercise";
import {getDateFormatYMD, getDateRelatedToThisDate} from "../../../helpers/calendar-helper";

let testRenderer;

beforeAll(() => {
  testRenderer = TestRenderer.create(
    <Router>
      <Exercise
        match={{
          params: {
            scenarioId: 11,
            date: getDateFormatYMD(getDateRelatedToThisDate(0, -4))
          }
        }}
      />
    </Router>
  );
});

test('renders correctly', () => {
  const tree = testRenderer.toJSON();
  expect(tree).toMatchSnapshot();
});
