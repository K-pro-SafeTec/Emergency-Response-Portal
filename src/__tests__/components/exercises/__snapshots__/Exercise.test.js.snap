// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <header
    className="MuiPaper-root-14 MuiPaper-elevation4-20 MuiAppBar-root-5 MuiAppBar-positionFixed-6 MuiAppBar-colorDefault-11 mui-fixed"
  >
    <div
      className="MuiToolbar-root-41 MuiToolbar-regular-43 MuiToolbar-gutters-42"
    >
      <a
        className="MuiButtonBase-root-51 MuiIconButton-root-45 AppPage-homeButton-2"
        href="/"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiIconButton-label-50"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-54"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-127"
        />
      </a>
      <button
        className="MuiButtonBase-root-51 MuiButtonBase-disabled-52 MuiIconButton-root-45 MuiIconButton-disabled-49"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
        type="button"
      >
        <span
          className="MuiIconButton-label-50"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-54"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root-51 MuiButtonBase-disabled-52 MuiIconButton-root-45 MuiIconButton-disabled-49"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
        type="button"
      >
        <span
          className="MuiIconButton-label-50"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-54"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
            />
          </svg>
        </span>
      </button>
      <h2
        className="MuiTypography-root-63 MuiTypography-title-69 MuiTypography-colorInherit-92 AppPage-grow-1"
      >
        Øvelser
      </h2>
      <a
        className="AppPage-safetecLink-3"
        href="https://www.safetec.no/"
      >
        <img
          alt="Levert av Safetec"
          height="44px"
          src="safetec.svg"
          width="153px"
        />
      </a>
    </div>
  </header>,
  <div
    className="AppPage-content-4"
  >
    <div
      className="MuiPaper-root-14 MuiPaper-elevation2-18 MuiPaper-rounded-15"
      style={
        Object {
          "margin": "1em",
          "padding": "1em",
        }
      }
    >
      <div
        className="EntityInfo-container-98"
      >
        <img
          alt="Øvelse"
          className="EntityInfo-avatar-99"
          src="exerciseScenarios.svg"
        />
        <div
          className="EntityInfo-info-100"
        >
          <h1
            className="MuiTypography-root-63 MuiTypography-headline-68"
          >
            Alvorlig syk/skadet person
          </h1>
          <h3
            className="MuiTypography-root-63 MuiTypography-subheading-70"
          >
            DFU 
            11
          </h3>
        </div>
      </div>
      <h3
        className="MuiTypography-root-63 MuiTypography-subheading-70"
      >
        Scenario:
      </h3>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72 MuiTypography-paragraph-91"
      >
        Personskade
      </p>
      <h3
        className="MuiTypography-root-63 MuiTypography-subheading-70"
      >
        Scenariobeskrivelse:
      </h3>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72 MuiTypography-paragraph-91"
      >
        Klemskade som følge av løfteoperasjon. Kontainer svinger ukontrollert og treffer en dekksarbeider. Personen er bevistløs og har synlige blødninger.
      </p>
      <h3
        className="MuiTypography-root-63 MuiTypography-subheading-70"
      >
        Forberedelser:
      </h3>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Sminke  markør
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Avtale med kranfører og forberede et skadested
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Avtale med dekksmannskap om varsling til kontrollrom, husk "Øvelse-øvelse-øvelse"
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Om mulig, simuler at markør ligger fast og må frigjøres
      </p>
      <h3
        className="MuiTypography-root-63 MuiTypography-subheading-70"
        style={
          Object {
            "marginTop": "12px",
          }
        }
      >
        Ytelseskrav:
      </h3>
      <h3
        className="MuiTypography-root-63 MuiTypography-subheading-70"
      >
        Fokusområder:
      </h3>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Livreddende stabilisering på skadested
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Koordinere frigjøring av forulykket med S&R lag
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Skånsom transport til hospital
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Forberede for MEDEVAC 
      </p>
      <h3
        className="MuiTypography-root-63 MuiTypography-subheading-70"
        style={
          Object {
            "marginTop": "12px",
          }
        }
      >
        Utvikling:
      </h3>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Potensiell alvorlig skade på forulykket, herunder brudd og indre blødninger
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Situasjonen forverrer seg grunnet indre blødning
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Sykepleier må følge helikopter til land med pasient
      </p>
      <p
        className="MuiTypography-root-63 MuiTypography-body1-72"
      >
        Tiltak på innretningen når sykepleier ikke lenger er tilstede
      </p>
      <a
        className="MuiButtonBase-root-51 MuiButton-root-101 MuiButton-contained-112 MuiButton-containedPrimary-113 MuiButton-raised-115 MuiButton-raisedPrimary-116"
        href="/undefineddebrief/"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        style={
          Object {
            "marginTop": "12px",
          }
        }
        tabIndex="0"
      >
        <span
          className="MuiButton-label-102"
        >
          Debrief
        </span>
        <span
          className="MuiTouchRipple-root-127"
        />
      </a>
    </div>
  </div>,
]
`;
